### PR TITLE
Enable pipeline to sync PSGallery modules to AzArtifacts feed

### PR DESCRIPTION
**Pull request by adityapatwardhan**
## PR Summary

Compare the PowerShell modules dependencies on PowerShell Gallery and sync them to AzDevOps artifacts feed if a newer is available. The release builds pick up modules from AzDevOps feed.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by adityapatwardhan** [comment]
I have tested this build in a private branch on AzDevOps.

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/PSGalleryToAzArtifacts.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    displayName: Download packages from PSGallery that need to be updated
+    condition: succeededOrFailed()
+
+  - powershell: |
+      Write-Verbose -Verbose "Packages to upload"
+      (Get-ChildItem "$(Build.ArtifactStagingDirectory)/*.nupkg").FullName
+    displayName: List packages to upload
+    condition: succeededOrFailed()
+
+  - task: NuGetCommand@2
```
Comment (anmenaga) [inline tools/releaseBuild/azureDevOps/AzArtifactFeed/PSGalleryToAzArtifacts.yml]: As we discussed; this needs to be double-checked.

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1 [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            $modulesToUpdate += $foundPackageOnGallery
+        } elseif ($foundPackageOnGallery.Version -lt $foundPackageOnAz.Version) {
+            Write-Warning "Newer version found on Az Artifacts - $($foundPackageOnAz.Name) - $($foundPackageOnAz.Version)"
+        } else {
+            Write-Verbose -Verbose "Module is in sync - $($p.Name)"
+        }
+    }
+
+    Write-Verbose -Verbose "Gallery Packages:"
+    $galleryPackages
```
Comment (anmenaga) [inline tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1]: IMHO either previous line `Gallery Packages:` should go into output stream;
or this line `$galleryPackages` should go into Verbose stream.
Similar for 2 instances below.

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/PSGalleryToAzArtifacts.yml (hunk 1 above)
Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/AzArtifactFeed/PSGalleryToAzArtifacts.yml]: Updated to use the PAT

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1 (hunk 2 above)
Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1]: Updated.

**Comment by adityapatwardhan** [comment]
@anmenaga @TravisEz13  Please have a look.

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1 [hunk 3]
```diff
@@ -0,0 +1,103 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+function SyncGalleryToAzArtifacts {
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1]: Describe the parameters

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1 [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    $galleryPackages = @()
+    $azArtifactsPackages = @()
+    $modulesToUpdate = @()
+
+    $galleryUrl = 'https://www.powershellgallery.com/api/v2/'
+    $azArtifactsUrl = 'https://mscodehub.pkgs.visualstudio.com/_packaging/pscore-release/nuget/v2'
+
+    $azDevOpsCreds = [pscredential]::new($env:AzDevOpsUserName, (ConvertTo-SecureString -String $AzDevOpsPAT -AsPlainText -Force))
+
+    foreach ($p in $packages) {
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1]: Use a better name than `$p`

**Review by anmenaga — APPROVED** [review]

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1 (hunk 3 above)
Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1]: Added help.

File: tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1 (hunk 4 above)
Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/AzArtifactFeed/SyncGalleryToAzArtifacts.psm1]: Fixed

**Comment by adityapatwardhan** [comment]
@TravisEz13 updated, please have a look.

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**